File: python/online_shopping_system.py
from abc import ABC, abstractmethod


# ---------------------------
# Abstraction
# ---------------------------
class User(ABC):
    def __init__(self):
        self._id = None
        self._email = None

    @abstractmethod
    def login(self):
        pass

    def register(self):
        print("User Registered Successfully.")


# ---------------------------
# Inheritance
# Customer inherits User
# ---------------------------
class Customer(User):
    def __init__(self):
        super().__init__()
        # Encapsulation
        self.__name = None
        self.__phone_number = None

    # Getter
    def get_name(self):
        return self.__name

    # Setter
    def set_name(self, name):
        self.__name = name

    # Getter
    def get_phone_number(self):
        return self.__phone_number

    # Setter
    def set_phone_number(self, phone_number):
        self.__phone_number = phone_number

    # Polymorphism (Method Overriding)
    def login(self):
        print("Customer Logged In Successfully.")


# ---------------------------
# Another child class
# ---------------------------
class Admin(User):
    def login(self):
        print("Admin Logged In Successfully.")


# ---------------------------
# Product Class
# ---------------------------
class Product:
    def __init__(self):
        self.product_id = None
        self.product_name = None
        self.price = None
        self.description = None
        self.stock = None

    def display_product(self):
        print("Displaying Product Details...")


# ---------------------------
# Product Category
# ---------------------------
class ProductCategory:
    def __init__(self):
        self.category_id = None
        self.category_name = None


# ---------------------------
# Order Class
# ---------------------------
class Order:
    def __init__(self):
        self.order_id = None
        self.order_date = None
        self.amount = None

    def place_order(self):
        print("Order Placed Successfully.")


# ---------------------------
# Payment Class
# ---------------------------
class Payment:
    # Polymorphism (Method Overloading, simulated via default args)
    def make_payment(self, amount, payment_method=None):
        if payment_method is None:
            print(f"Paid ₹{amount} using Default Method.")
        else:
            print(f"Paid ₹{amount} using {payment_method}.")


# ---------------------------
# Wishlist
# ---------------------------
class Wishlist:
    def add_product(self):
        print("Product Added to Wishlist.")

    def remove_product(self):
        print("Product Removed from Wishlist.")


# ---------------------------
# Search
# ---------------------------
class Search:
    def search_product(self, keyword):
        print(f"Searching for: {keyword}")


# ---------------------------
# Offer
# ---------------------------
class Offer:
    def __init__(self):
        self.coupon = None
        self.discount = None


# ---------------------------
# Delivery
# ---------------------------
class Delivery:
    def __init__(self):
        self.delivery_id = None
        self.delivery_status = None

    def track_order(self):
        print("Your Order is Out for Delivery.")


# ---------------------------
# Review
# ---------------------------
class Review:
    def __init__(self):
        self.rating = None
        self.feedback = None


# ---------------------------
# Profile
# ---------------------------
class Profile:
    def __init__(self):
        self.address = None
        self.username = None

    def update_profile(self):
        print("Profile Updated Successfully.")


# ---------------------------
# Main
# ---------------------------
def main():
    print("====== ONLINE SHOPPING SYSTEM ======\n")

    # Abstraction + Inheritance
    customer = Customer()
    customer.register()
    customer.login()

    # Encapsulation
    customer.set_name("Rakshita")
    customer.set_phone_number(9876543210)

    print(f"Customer Name : {customer.get_name()}")
    print(f"Phone Number  : {customer.get_phone_number()}")

    print()

    # Another inherited class
    admin = Admin()
    admin.register()
    admin.login()

    print()

    # Product
    product = Product()
    product.product_name = "Laptop"
    product.price = 55000

    product.display_product()

    print(f"Product : {product.product_name}")
    print(f"Price   : ₹{product.price}")

    print()

    # Order
    order = Order()
    order.order_id = 101
    order.order_date = "06-Aug-2026"
    order.amount = 55000

    order.place_order()

    print()

    # Polymorphism (Method Overloading)
    payment = Payment()
    payment.make_payment(55000)
    payment.make_payment(55000, "UPI")

    print()

    # Wishlist
    wishlist = Wishlist()
    wishlist.add_product()

    # Search
    search = Search()
    search.search_product("Laptop")

    # Delivery
    delivery = Delivery()
    delivery.track_order()

    # Profile
    profile = Profile()
    profile.update_profile()

    print("\nOnline Shopping System Executed Successfully.")


if __name__ == "__main__":
    main()

File: c#/OnlineShoppingSystem.cs
using System;

namespace OnlineShopping
{
    // ---------------------------
    // Abstraction
    // ---------------------------
    abstract class User
    {
        protected int id;
        protected string email;

        // Abstract method (no implementation)
        public abstract void Login();

        // Normal method
        public void Register()
        {
            Console.WriteLine("User Registered Successfully.");
        }
    }

    // ---------------------------
    // Inheritance
    // Customer inherits User
    // ---------------------------
    class Customer : User
    {
        // Encapsulation
        private string name;
        private long phoneNumber;

        // Property (Getter/Setter)
        public string Name
        {
            get { return name; }
            set { name = value; }
        }

        public long PhoneNumber
        {
            get { return phoneNumber; }
            set { phoneNumber = value; }
        }

        // Polymorphism (Method Overriding)
        public override void Login()
        {
            Console.WriteLine("Customer Logged In Successfully.");
        }
    }

    // ---------------------------
    // Another child class
    // ---------------------------
    class Admin : User
    {
        public override void Login()
        {
            Console.WriteLine("Admin Logged In Successfully.");
        }
    }

    // ---------------------------
    // Product Class
    // ---------------------------
    class Product
    {
        public int productId;
        public string productName;
        public double price;
        public string description;
        public int stock;

        public void DisplayProduct()
        {
            Console.WriteLine("Displaying Product Details...");
        }
    }

    // ---------------------------
    // Product Category
    // ---------------------------
    class ProductCategory
    {
        public int categoryId;
        public string categoryName;
    }

    // ---------------------------
    // Order Class
    // ---------------------------
    class Order
    {
        public int orderId;
        public string orderDate;
        public double amount;

        public void PlaceOrder()
        {
            Console.WriteLine("Order Placed Successfully.");
        }
    }

    // ---------------------------
    // Payment Class
    // ---------------------------
    class Payment
    {
        // Polymorphism (Method Overloading)

        public void MakePayment(double amount)
        {
            Console.WriteLine("Paid ₹" + amount + " using Default Method.");
        }

        public void MakePayment(double amount, string paymentMethod)
        {
            Console.WriteLine("Paid ₹" + amount + " using " + paymentMethod + ".");
        }
    }

    // ---------------------------
    // Wishlist
    // ---------------------------
    class Wishlist
    {
        public void AddProduct()
        {
            Console.WriteLine("Product Added to Wishlist.");
        }

        public void RemoveProduct()
        {
            Console.WriteLine("Product Removed from Wishlist.");
        }
    }

    // ---------------------------
    // Search
    // ---------------------------
    class Search
    {
        public void SearchProduct(string keyword)
        {
            Console.WriteLine("Searching for: " + keyword);
        }
    }

    // ---------------------------
    // Offer
    // ---------------------------
    class Offer
    {
        public string coupon;
        public double discount;
    }

    // ---------------------------
    // Delivery
    // ---------------------------
    class Delivery
    {
        public int deliveryId;
        public string deliveryStatus;

        public void TrackOrder()
        {
            Console.WriteLine("Your Order is Out for Delivery.");
        }
    }

    // ---------------------------
    // Review
    // ---------------------------
    class Review
    {
        public int rating;
        public string feedback;
    }

    // ---------------------------
    // Profile
    // ---------------------------
    class Profile
    {
        public string address;
        public string username;

        public void UpdateProfile()
        {
            Console.WriteLine("Profile Updated Successfully.");
        }
    }

    // ---------------------------
    // Main Class
    // ---------------------------
    class OnlineShoppingSystem
    {
        static void Main(string[] args)
        {
            Console.WriteLine("====== ONLINE SHOPPING SYSTEM ======\n");

            // Abstraction + Inheritance
            Customer customer = new Customer();
            customer.Register();
            customer.Login();

            // Encapsulation
            customer.Name = "Rakshita";
            customer.PhoneNumber = 9876543210L;

            Console.WriteLine("Customer Name : " + customer.Name);
            Console.WriteLine("Phone Number  : " + customer.PhoneNumber);

            Console.WriteLine();

            // Another inherited class
            Admin admin = new Admin();
            admin.Register();
            admin.Login();

            Console.WriteLine();

            // Product
            Product product = new Product();
            product.productName = "Laptop";
            product.price = 55000;

            product.DisplayProduct();

            Console.WriteLine("Product : " + product.productName);
            Console.WriteLine("Price   : ₹" + product.price);

            Console.WriteLine();

            // Order
            Order order = new Order();
            order.orderId = 101;
            order.orderDate = "06-Aug-2026";
            order.amount = 55000;

            order.PlaceOrder();

            Console.WriteLine();

            // Polymorphism (Method Overloading)
            Payment payment = new Payment();
            payment.MakePayment(55000);
            payment.MakePayment(55000, "UPI");

            Console.WriteLine();

            // Wishlist
            Wishlist wishlist = new Wishlist();
            wishlist.AddProduct();

            // Search
            Search search = new Search();
            search.SearchProduct("Laptop");

            // Delivery
            Delivery delivery = new Delivery();
            delivery.TrackOrder();

            // Profile
            Profile profile = new Profile();
            profile.UpdateProfile();

            Console.WriteLine("\nOnline Shopping System Executed Successfully.");
        }
    }
}
